src/full/Agda/Syntax/Concrete/Definitions.hs

```diff
@@ -1513,8 +776,189 @@ niceDeclarations fixs ds = do
                not (null notStrings) && (notStrings `isSublistOf` patStrings)
         -- not a notation, not first id: give up
         _ -> False -- trace ("couldBe not (case default)") $ False
+
+
+    -- for finding nice clauses for a type sig in mutual blocks
+    couldBeNiceFunClauseOf :: Maybe Fixity' -> Name -> NiceDeclaration
+                           -> Maybe (MutualChecks, Declaration)
+    couldBeNiceFunClauseOf mf n (NiceFunClause _ _ _ tc cc _ d)
+      = (MutualChecks [tc] [cc] [], d) <$ guard (couldBeFunClauseOf mf n d)
+    couldBeNiceFunClauseOf _ _ _ = Nothing
+
+    -- for finding clauses for a type sig in mutual blocks
+    couldBeFunClauseOf :: Maybe Fixity' -> Name -> Declaration -> Bool
+    couldBeFunClauseOf mFixity x (Pragma (CatchallPragma{})) = True
+    couldBeFunClauseOf mFixity x (FunClause (LHS p _ _ _) _ _ _) =
+       hasEllipsis p || couldBeCallOf mFixity x p
     couldBeFunClauseOf _ _ _ = False -- trace ("couldBe not (fun default)") $ False
 
+    -- | Turn an new style interleaving mutual block into a new style mutual block
+    --   by grouping the declarations in blocks.
+    mkInfixMutual
+      :: Range                -- ^ Range of the whole @mutual@ block.
+      -> [NiceDeclaration]    -- ^ Declarations inside the block.
+      -> Nice NiceDeclaration -- ^ Returns a 'NiceMutual'.
+    mkInfixMutual r ds' = do
+      (other, (m, checks, _)) <- runStateT (groupByBlocks r ds') (empty, mempty, 0)
+      let dec k (InfixData (i, d@(NiceDataSig _ acc abs pc uc _ pars _)) ds) =
```

Comment:
TODO: move these transformations next to the `InfixData` / `InfixFun` introduction site.